### ISSUE TITLE
unit tests: pmem_is_pmem/TEST1 fails on pmem(DAX)

### ISSUE THREAD

**Issue by kkajrewicz** [pmem/issues#207]
```
pmem_is_pmem/TEST1: SETUP (check/pmem/debug)
pmem_is_pmem/TEST1: START: pmem_is_pmem
pmem_is_pmem/TEST1 crashed (signal 6). err1.log below.
{pmem_is_pmem.c:91 main} pmem_is_pmem/TEST1: Error: assertion failure: ret[0] (0x1) == ret[i] (0x0)

Last 30 lines of pmem1.log below (whole file has 32 lines).
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <1> [out.c:239 out_init] src version SRCVERSION:1.1-70-g594bb38
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [libpmem.c:57 libpmem_init] 
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [util.c:96 util_init] 
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [pmem.c:1161 pmem_init] 
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [pmem.c:1099 pmem_get_cpuinfo] clflush supported
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [pmem.c:1103 pmem_get_cpuinfo] clflushopt supported
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [pmem.c:1115 pmem_get_cpuinfo] clwb supported
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [pmem.c:1127 pmem_get_cpuinfo] using clwb
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [pmem.c:1136 pmem_get_cpuinfo] movnt supported
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [pmem.c:1148 pmem_get_cpuinfo] using movnt
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [pmem.c:454 pmem_is_pmem_init] 
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [pmem.c:454 pmem_is_pmem_init] 
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [pmem.c:454 pmem_is_pmem_init] 
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [pmem_linux.c:135 is_pmem_proc] returning 1
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [pmem_linux.c:135 is_pmem_proc] returning 1
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [pmem_linux.c:135 is_pmem_proc] returning 1
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [pmem_linux.c:135 is_pmem_proc] returning 1
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [pmem_linux.c:135 is_pmem_proc] returning 1
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [pmem_linux.c:135 is_pmem_proc] returning 1
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [pmem_linux.c:135 is_pmem_proc] returning 1
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [pmem_linux.c:135 is_pmem_proc] returning 1
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [pmem_linux.c:135 is_pmem_proc] returning 0
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [pmem_linux.c:135 is_pmem_proc] returning 1
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [pmem_linux.c:135 is_pmem_proc] returning 1
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [pmem_linux.c:135 is_pmem_proc] returning 1
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [pmem_linux.c:135 is_pmem_proc] returning 1
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [pmem_linux.c:135 is_pmem_proc] returning 1
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [pmem_linux.c:135 is_pmem_proc] returning 0
pmem_is_pmem/TEST1 pmem1.log <libpmem>: <3> [pmem_linux.c:135 is_pmem_proc] returning 1
```

Found in 1.1-70-g594bb38

**Comment by ryspol** [comment]
Test calls is_pmem_proc() which parses /proc/self/smaps. Looks that the reason is lack of needed memory region in that file when many threads are activated. Behavior maybe connected to the internal mechanism of updating process' memory in smaps file by kernel - memory region is not visible yet in file due to some synchronization problem.

**Comment by marcinslusarz** [comment]
I reproduced this problem with anonymous mappings.
I reported it to Linux Kernel Mailing List: https://lkml.org/lkml/2016/7/26/241
The current thinking is that /proc/self/smaps is unfixable.

However I investigated this problem on the kernel side and I think this particular case should be fixable, but I haven't pinpointed where the bug exactly is. For now I don't have time to look into this, so we'll need to disable multi-threaded pmem_is_pmem test or change pmem_is_pmem_proc to loop on reading /proc/self/smaps until we find the mapping. The disadvantage is of course an infinite loop if user passes invalid address to pmem_is_pmem (documentation fix?).

**Comment by LingyanYin** [comment]
Is this bug fixed? I came across the same issue while I run unit tests in my computer.

**Comment by krzycz** [comment]
No, it's not fixed yet, but we already have an idea how to workaround this issue (w/o the disadvantage described above), so the fix/workaround should be merged soon.

**Comment by marcinslusarz** [comment]
FTR, this bug was fixed in Linux 4.9:
https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=855af072b6c40aeb266f4dc98fd9a6a49edf22af

**Comment by plebioda** [comment]
We don't use smaps anymore so I'm closing this issue.
